### ISSUE TITLE
Subscriptions do not maintain ordering

### ISSUE THREAD

**Issue by jkxyz** [walmartlabs/lacinia#460]
I've been using Lacinia subscriptions to handle AWS event stream responses, to stream those events back to the client. 

It seems that the values passed to the subscription callback function are not sent to the WebSocket in the same order that the callback is called. I can log the order on the server, and then by observing the WebSocket messages I can see that they're being sent in a different order.

Here's a stripped down version of the subscription handler:

```clj
(defn subscription-handler [_context _args callback]
  (let [cancel-ch (async/chan)
        events-ch (async/chan)

        fut
        (future
          (invoke-model-stream events-ch ,,,))]
    (async/go-loop []
      (async/alt!
        cancel-ch nil
        events-ch ([event]
                   (callback event)
                   (when event
                     (recur)))))
    (fn []
      (async/close! cancel-ch)
      @fut)))
```

It doesn't seem to be documented anywhere that Lacinia subscriptions don't provide ordering guarantees. For now I've managed to work around this by providing an index in the responses and re-ordering on the client.

**Comment by hlship** [comment]
I'm surprised that order is not maintained, I'll have to look over the code and the guarantees made by core.async.  How close are the events from AWS coming in?

**Comment by jkxyz** [comment]
> How close are the events from AWS coming in?

I've been testing this with the InvokeModelWithResponseStream operation on the AWS Bedrock API. So the events are coming in milliseconds apart. The rate is not constant and most events do arrive on the client in the correct order. 

I can provide more surrounding code to try and get a minimal reproduction, if it helps.

**Comment by hlship** [comment]
I'm putting this on a milestone just to keep track of it, I haven't looked at it at all (to verify, to reproduce it, etc.).

**Comment by hlship** [comment]
I'm not sure what's going on here; from your sketch, it looks like there's only a single writer to the channel, and order is (AFAIK) guaranteed for writes to a channel, so I'm not aware of anything that could affect the ordering of data read from the channel(s) and written to the client.

**Comment by jkxyz** [comment]
Hi Howard - I'm no longer working at Nosco so it's not easy for me to provide more context. I spent half a day testing this and in the end it was a blocker for us using subscriptions on that feature. But I couldn't find any obvious reasons for the bug. Thanks for your work on Lacinia!

**Comment by ebedictus** [comment]
I suspect the problem lies in `lacinia-pedestal`'s `execute-subscription`, which does not seem to maintain insertion order for a single subscription. Correct me if I'm wrong, but when `value`s come from `source-stream-ch`, the resolver execution tasks for each `value` are sent to `lacinia`'s thread pool executor by `executor/execute-query`.  Because the executor has ten threads by default, _some of these tasks may execute in parallel, causing their results to be pushed to the `response-data-ch` in an arbitrary order._ While this is perfectly fine for tasks from different (independent) subscriptions, tasks coming from a single subscription should probably be executed serially. This would ensure that results are pushed to `response-data-ch` in insertion order.

Additionally, this thread macro parallelizes the executions as well, meaning that fixing the thread pool issue alone does not seem to solve the problem: https://github.com/walmartlabs/lacinia-pedestal/blob/e13ee078be0f3a86df817f94cac3665aed5e8572/src/com/walmartlabs/lacinia/pedestal/subscriptions.clj#L393

I experimented by using a single-thread executor and removing the thread macro from `execute-subscription`. After making these modifications, the insertion order was preserved. However, I am not sure how to fix this properly, i.e., to use the thread pool for independent subscriptions only. And thanks for your work on Lacinia. Unfortunately, this is a blocker for us as well

**Comment by hlship** [comment]
Thanks so much for the research! I can certainly look into this in more detail, documenting the current behavior but also looking to add an option for more rigorous ordering.

**Comment by hlship** [comment]
I'm refamiliarizing myself with the code. 

I'm not sure what ordering guarantees I can make.  GraphQL queries (as opposed to mutations) are designed to be parallelizable, and the applications is supposed to ensure idempotence. What you are proposing is a mode where each async query (from the WS client) must be processed and produce a result (streamed back to the client) before any further async queries may be processed.  I'm looking into whether there can be some support for this in the lacinia-pedestal layer, but I'm thinking this belongs more on the application side.

**Comment by hlship** [comment]
The research I've done does indicate that the GraphQL spec does not mandate any ordering for subscriptions. I think the application could provide better ordering, but there are limitations and uncertainties; I would suggest changing your streamer to make use of a queue (maybe a core async channel) to serialize the processing of each incoming event, so that they are processed in a deterministic order.  Relying on thread pools and `future` will never have an assurance of ordering.

**Comment by sami-silvennoinen** [comment]
The spec says: “Executing a subscription operation creates a persistent function on the service that maps an underlying source stream to a returned response stream.” (https://spec.graphql.org/October2021/#sec-Subscription).

The response stream is therefore defined as a **mapping** of the source stream. A mapping preserves the sequence of elements unless the specification explicitly authorises reordering. 

Lacinia:
- Receives source events in order.
- Starts execution of each event in separate threads.
- Emits results as soon as execution completes.

This is no longer the stream obtained by applying MapSourceToResponseEvent to the source stream. It is a different stream whose ordering depends on execution latency rather than source-event order.

Furthermore, the spec algorithm says that when sourceStream emits sourceValue, execute it and emit that result on responseStream (https://spec.graphql.org/draft/#sec-Subscription.Response-Stream).

It does **not** say: “Start executions concurrently and emit results in whatever order they finish.”

An implementation may optimise internally, but it must produce an **observationally equivalent result**. Returning events in completion order is observable by the client, so it changes the semantics.

Also, the graphql-js **reference implementation** retains the ordering. It handles the source stream events in a single for await...of loop exactly according to the spec (http://github.com/graphql/graphql-js/blob/%E2%80%A6/mapAsyncIterable.ts, and http://github.com/graphql/graphql-js/blob/%E2%80%A6/execute.ts).

If the order is not preserved, subscriptions could not be used for most stateful applications where the order matters, such as:

- chat messages (also given as an example application in the spec, see https://spec.graphql.org/draft/#sel-GANTJLABaBe8pP)
- audit logs
- financial transactions
- sequence-numbered domain events

**Comment by sami-silvennoinen** [comment]
Reactive UIs become significantly harder to implement because a later state update can arrive before an earlier one. A UI element whose state is driven by subscriptions may temporarily or even permanently display stale state unless additional versioning, sequencing, or reconciliation logic is added by the client. The classic example is a toggle/button that is switched on and then immediately off; if the subscription updates are reordered, the UI can incorrectly end up showing "on" while the actual backend state is already "off".

**Comment by jkxyz** [comment]
> If the order is not preserved, subscriptions could not be used for most stateful applications where the order matters

They can be used for many applications, but it requires the stream elements to contain an index allowing the client to buffer any out-of-order elements it receives.

**Comment by sami-silvennoinen** [comment]
> > If the order is not preserved, subscriptions could not be used for most stateful applications where the order matters
> 
> They can be used for many applications, but it requires the stream elements to contain an index allowing the client to buffer any out-of-order elements it receives.

Yes, but it requires the same buffering feature in _all the clients_. It would make more sense if the server did it once and for all, instead of each client implementing it separately.

**Comment by terhohautala** [comment]
@hlship, I think that your suggestion of 
> ...changing your streamer to make use of a queue […] processed in a deterministic order.

wouldn’t work, because even if the streamer is emitting the events in deterministic order, Lacinia internally runs the resolvers for each consecutive input in separate threads which emit the results in completion order (so the streamer’s ordering is not maintained). Or did I misunderstand your suggestion? Can you please elucidate how that would work?
